### ISSUE TITLE
limit cpus seen by hwloc in Flux jobs

### ISSUE THREAD

**Issue by ryanday36** [flux-framework/flux-core#5934]
Is it possible to limit the cpus seen by hwloc to just those assigned to a Flux job? e.g., on fluke if I run with `flux run -n2 -o mpibind=off ...` The job gets two cpus, and I only see those with `numactl -s`, but `lstopo-no-graphics` still lists all four cpus:

```
[day36@fluke95:~]$ flux run -n2 -l -o mpibind=off bash -c "numactl -s | grep -i physcpu && lstopo-no-graphics | grep L2"
0: physcpubind: 2 3 6 7 
1: physcpubind: 2 3 6 7 
0: hwloc/rsmi: Failed to initialize with rsmi_init(): RSMI_STATUS_INIT_ERROR: An error occurred during initialization, during monitor discovery or when when initializing internal data structures
0:       L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
0:       L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
0:       L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
0:       L2 L#3 (256KB) + L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3
1: hwloc/rsmi: Failed to initialize with rsmi_init(): RSMI_STATUS_INIT_ERROR: An error occurred during initialization, during monitor discovery or when when initializing internal data structures
1:       L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
1:       L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
1:       L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
1:       L2 L#3 (256KB) + L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3
```

I also tried running with mpibind and setting `-o cpu-affinity=on` and `-o hwloc.xmlfile` with the same results (excepting that you have to `grep Core` rather than `L2` with `-o hwloc.xmlfile`).

For background, this came up because a user is trying to run two jobs on the same node with a multi-physics code that does its own task and thread affinity using hwloc and both jobs appear to be trying to use all of the cpus on the node.

**Comment by grondo** [comment]
It works if you add the `--restrict binding` flag to `lstopo-no-graphics`. I'll dig in and see what flag that uses during `hwloc_topology_load()`. 

```
f(s=1,d=1) grondo@fluke94:~/git/flux-core.git$ flux run -n2 -l -o mpibind=off bash -c "numactl -s | grep -i physcpu && lstopo-no-graphics --restrict binding| grep L2"
0: physcpubind: 2 3 6 7 
1: physcpubind: 2 3 6 7 
1: hwloc/rsmi: Failed to initialize with rsmi_init(): RSMI_STATUS_INIT_ERROR: An error occurred during initialization, during monitor discovery or when when initializing internal data structures
0: hwloc/rsmi: Failed to initialize with rsmi_init(): RSMI_STATUS_INIT_ERROR: An error occurred during initialization, during monitor discovery or when when initializing internal data structures
1:       L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
1:       L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
0:       L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
0:       L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
```

I'm surprised and disheartened that the `hwloc.xmlfile` option doesn't work. This makes me think hwloc or the utilities are ignoring `HWLOC_XMLFILE` by default, so this needs to be investigated.

**Comment by wihobbs** [comment]
If I'm understanding you correctly, you _may_ be able to use the `lstopo --restrict` option to pass a cpuset to hwloc that restricts what it can discover like so:

```
(s=1,d=1)  corona276 ~ $ flux resource list
     STATE PROPERTIES NNODES   NCORES    NGPUS NODELIST
      free pbatch          1       48        8 corona276
 allocated                 0        0        0
      down                 0        0        0
(s=1,d=1)  corona276 ~ $ lstopo --restrict 0x1,0x2,0x3,0x4 --of xml > /tmp/file.xml
(s=1,d=1)  corona276 ~ $ FLUX_HWLOC_XMLFILE=/tmp/file.xml flux start flux resource list
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1        4        8 corona276
 allocated      0        0        0
      down      0        0        0
```
I put the `flux start flux resource list` call in there just to make sure the hwloc XML file output was what I thought it was. Maybe pass that xml file to the users' code doing the mapping, or set `HWLOC_XMLFILE` in the environment to restrict its' discovery?

**Comment by wihobbs** [comment]
Oh, heh. @grondo beat me!

**Comment by grondo** [comment]
> or set HWLOC_XMLFILE in the environment to restrict its' discovery?

This is what `-o hwloc.xmlfile` does -- it should set `HWLOC_XMLFILE` to the hwloc XML fetched from the shell, which should already be restricted to the allocated resources.

**Comment by grondo** [comment]
Ah, I may have been mistaken in thinking the hwloc XML held by the shell and thus written out via the `hwloc.xmlfile` option is restricted in any way. AFAICT, the shell fetches the copy from the enclosing instance and holds on to it unmodified for use in shell plugins.

**Comment by grondo** [comment]
Checking out the hwloc sources, this is how `hwloc-info` handles `--restrict binding`:
(Yes, they do use a mix of tabs and spaces in the source, making the conditionals "fun" to read)
```c
   if (!strcmp (restrictstring, "binding")) {
      if (pid_number > 0)
    hwloc_get_proc_cpubind(topology, pid, restrictset, HWLOC_CPUBIND_PROCESS);
      else
    hwloc_get_cpubind(topology, restrictset, HWLOC_CPUBIND_PROCESS);
    } else {
      hwloc_bitmap_sscanf(restrictset, restrictstring);
    }
    err = hwloc_topology_restrict (topology, restrictset, restrict_flags);
    if (err) {
      perror("Restricting the topology");
      /* FALLTHRU */
    }
```

The user may have to add this to their code to restrict the topology to the current binding.

Alternatively, I wonder if we can add a mode to `hwloc.xmlfile` that restricts the topology on behalf of the user, and maybe that would "just work"?

**Comment by ryanday36** [comment]
> It works if you add the `--restrict binding` flag to `lstopo-no-graphics`. I'll dig in and see what flag that uses during `hwloc_topology_load()`.

Thanks @grondo! Knowing what `--restrict binding` is doing under the covers would definitely be useful as that would allow the developers for the package to restrict themselves too.

Slurm restricts what hwloc sees via, I believe, the job cgroup. Do you know if there's anything we can do with our cgroup configs to possibly limit this as well?

**Comment by grondo** [comment]
> Slurm restricts what hwloc sees via, I believe, the job cgroup. Do you know if there's anything we can do with our cgroup configs to possibly limit this as well?

Yeah, hwloc does appear to restrict to "allowed" resources by default, which includes restrictions by cgroup but not binding.

We would need to add AllowedCPUs to the cgroups we're creating with the transient system units. However, this would only work at the top level. More effort would be needed to manually create cgroups for subinstances under that to get the same effect all the way down. Unfortunately, I  don't think we've had time to see if that's even possible.

**Comment by grondo** [comment]
Here's a possible alternate solution. It was quite simple to add a `hwloc.restrict` flag which restricts the hwloc XML to the current process binding before writing it out to `HWLOC_XMLFILE`.

```
f(s=1,d=1,builddir) grondo@fluke95:~/git/flux-core.git$ flux run -o mpibind=off -n1 -o hwloc.xmlfile hwloc-info
depth 0:           1 Machine (type #0)
 depth 1:          1 Package (type #1)
  depth 2:         4 Core (type #2)
   depth 3:        8 PU (type #3)
Special depth -3:  1 NUMANode (type #13)
Special depth -4:  5 Bridge (type #14)
Special depth -5:  5 PCIDev (type #15)
Special depth -6:  5 OSDev (type #16)
f(s=1,d=1,builddir) grondo@fluke95:~/git/flux-core.git$ flux run -o mpibind=off -n1 -o hwloc.xmlfile -o hwloc.restrict hwloc-info
depth 0:           1 Machine (type #0)
 depth 1:          1 Package (type #1)
  depth 2:         1 Core (type #2)
   depth 3:        2 PU (type #3)
Special depth -3:  1 NUMANode (type #13)
Special depth -4:  5 Bridge (type #14)
Special depth -5:  5 PCIDev (type #15)
Special depth -6:  5 OSDev (type #16)
```

**Comment by ryanday36** [comment]
Nice! That sounds quite promising.

**Comment by grondo** [comment]
Ok, I will propose a PR and get this feature into the next release in case it is useful.

One potential issue I noticed, however, is that the L2 caches seem to be dropped in our xmlfile output, at least from lstopo's perspective: (`lstopo.out` is just `lstopo-no-graphics` default output on a fluke node, while `xmlfile.out` is the output of `lstopo-no-graphics` when `HWLOC_XMLFILE` is set by the `hwloc.xmlfile` option)

```diff
--- lstopo.out	2024-05-06 09:17:27.181792000 -0700
+++ xmlfile.out	2024-05-06 09:17:19.595160000 -0700
@@ -1,19 +1,18 @@
 Machine (15GB total)
   Package L#0
     NUMANode L#0 (P#0 15GB)
-    L3 L#0 (8192KB)
-      L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
-        PU L#0 (P#0)
-        PU L#1 (P#4)
-      L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
-        PU L#2 (P#1)
-        PU L#3 (P#5)
-      L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
-        PU L#4 (P#2)
-        PU L#5 (P#6)
-      L2 L#3 (256KB) + L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3
-        PU L#6 (P#3)
-        PU L#7 (P#7)
+    Core L#0
+      PU L#0 (P#0)
+      PU L#1 (P#4)
+    Core L#1
+      PU L#2 (P#1)
+      PU L#3 (P#5)
+    Core L#2
+      PU L#4 (P#2)
+      PU L#5 (P#6)
+    Core L#3
+      PU L#6 (P#3)
+      PU L#7 (P#7)
   HostBridge
     PCIBridge
       PCI 01:00.0 (Network)
```

**Comment by grondo** [comment]
This difference seems to come from the following commented out lines in this diff. With this patch applied I get all the caches even with `-o hwloc.xmlfile`. 

```diff
diff --git a/src/common/librlist/rhwloc.c b/src/common/librlist/rhwloc.c
index 20f006e6f..4ea0ea832 100644
--- a/src/common/librlist/rhwloc.c
+++ b/src/common/librlist/rhwloc.c
@@ -40,10 +40,10 @@ static int topo_init_common (hwloc_topology_t *tp, unsigned long flags)
                                            HWLOC_TYPE_FILTER_KEEP_IMPORTANT)
         < 0)
         return -1;
-    if (hwloc_topology_set_cache_types_filter(*tp,
-                                              HWLOC_TYPE_FILTER_KEEP_STRUCTURE)
-        < 0)
-        return -1;
+    //if (hwloc_topology_set_cache_types_filter(*tp,
+    //                                          HWLOC_TYPE_FILTER_KEEP_STRUCTURE)
+    //    < 0)
+    //    return -1;
     if (hwloc_topology_set_icache_types_filter(*tp,
                                                HWLOC_TYPE_FILTER_KEEP_STRUCTURE)
         < 0)
```

I wonder if we should go ahead and make this change, but I don't remember why these filtering rules are being used in the first place.

@trws, do you recall? Is it now safe to remove at least this one filtering flag on caches? It might make the `-o hwloc.xmlfile` option more generally useful.

**Comment by grondo** [comment]
Good point. To reduce the use of `hwloc_topology_load()` (esp in parallel) we currently load hwloc topolog at the top level flux instance, then the job shell and subinstances alway pull the topology XML from the parent so all the way down the hierarchy the original XML is reused.

Currently there's no issue with cgroups because Flux doesn't support cgroups for CPUs or other devices discovered by hwloc. However, I wonder if the correct long term approach is to have this environment variable or the equivalent flag on by default so that each job shell and instance holds a restricted topology instead of the whole thing.

At this point, I don't know how this change would affect mpibind, so I'm hesitant to make the restricted XML the default. We can add the `hwloc.restrict` option for now, as well as adding back in the caches, and open a separate issue for the allowed resources investigation.

**Comment by trws** [comment]
Keeping the full xml contents but adding the `restrict` flag, and/or the env var, should play nice with mpibind.  It does its own either subsetting or not of the XML based on its own settings, so cutting it down to just what we're giving it might break other functionality some users rely on to subvert what they told us.

**Comment by grondo** [comment]
Ok, that is what I was thinking as well. We'll just keep an option so that restricting the XML is explicit.

**Comment by grondo** [comment]
Is the `hwloc.restrict` option enough to close this issue for now @ryanday36? I'm not sure we could make it the default due to @trws comment above.

**Comment by ryanday36** [comment]
> Is the `hwloc.restrict` option enough to close this issue for now @ryanday36? I'm not sure we could make it the default due to @trws comment above.

Yes. That looks like it'll do what the user wants. Thanks!
